File: engine-minicom/src/main/java/com/sos/scheduler/engine/minicom/idispatch/annotation/invocable.java
package com.sos.scheduler.engine.minicom.idispatch.annotation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Annotated method can be dynamically invoked via com.sos.scheduler.engine.minicom.idispatch.OverridingInvocableIDispatch#invoke.
 */
@Retention(RUNTIME)
@Target(ElementType.METHOD)
public @interface invocable {
    int UnusedDispId = Integer.MIN_VALUE;

    int dispId() default UnusedDispId;
}

File: minicom/src/main/scala/com/sos/scheduler/engine/minicom/idispatch/Invocable.scala
package com.sos.scheduler.engine.minicom.idispatch

import com.sos.scheduler.engine.minicom.types.IUnknown
import java.lang.reflect.Method
import scala.collection.immutable.Seq

/**
 * Public methods annotated with @[[com.sos.scheduler.engine.minicom.idispatch.annotation.invocable]] are callable via [[OverridingInvocableIDispatch]].
 * Nearly a [[IDispatch]], but without need to implement the `IDispatch`s methods.
 *
 * @author Joacim Zschimmer
 */
trait Invocable extends IUnknown {
  def invocableMethods: Seq[(Method, Option[DISPID])]
}

object Invocable {
  trait Empty extends Invocable {
    final def invocableMethods = Nil
  }
}
